### PR TITLE
EFR32 Window Covering sample added.

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
#### Problem
Missing EFR32 sample for the Window Covering cluster.

 #### Summary of Changes
The new EFR32 example supports all the Window Covering cluster commands. The Window Covering cluster has also been added to the supported clusters in controller-clusters.zap, so is now supported by the CHIP Device Controller.

### CODE REVIEW

**Comment by woody-apple** [comment]
@ricardo-casallas 
```
../../../../../config/esp32/third_party/connectedhomeip/src/app/tests/TestReadInteraction.cpp:151: assertion failed: "err == CHIP_NO_ERROR"
1981
```

Looks like tests are failing :(

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@ricardo-casallas can you please restyle?

@saurabhst ?

**Review by saurabhst — APPROVED** [review]

**Comment by jmeg-sfy** [comment]
@ricardo-casallas this PR need a restyling

File: examples/window-app/efr32/include/WindowCover.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ */
+
+#pragma once
+
+#include <AppEvent.h>
+#include <AppTimer.h>
+#include <memory>
+#include <stdint.h>
+#include <vector>
+// #include "Reporter.h"
```
Comment (mspang) [inline examples/window-app/efr32/include/WindowCover.h:26]: Can we remove this commented out code?

File: examples/window-app/common/window-app.zap [hunk 2]
```diff
@@ -0,0 +1,898 @@
...
+    },
+    {
+      "key": "manufacturerCodes",
+      "value": "0x1002"
+    }
+  ],
+  "package": [
+    {
+      "pathRelativity": "relativeToZap",
+      "path": "../../../../../../../../../git/open/connectedhomeip/src/app/zap-templates/zcl/zcl.json",
```
Comment (mspang) [inline examples/window-app/common/window-app.zap:21]: Are these the right paths? My repo is not in a directory called "git/open"

File: examples/window-app/efr32/include/WindowCover.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+// #include "AppTimer.h"
+
+class CoverReporter;
+
+class WindowCover
+{
+public:
+    typedef enum CoverType
```
Comment (mspang) [inline examples/window-app/efr32/include/WindowCover.h:34]: nit: Prefer `enum class`

File: examples/window-app/efr32/src/AppEvent.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <AppEvent.h>
+
+AppEvent::AppEvent() : mType(kEvent_None), mContext(nullptr) {}
```
Comment (mspang) [inline examples/window-app/efr32/src/AppEvent.cpp:22]: nit: Move these scalar initializers to the class definition, then specify this ctor as

AppEvent::AppEvent() = default;

File: examples/window-app/efr32/src/AppTimer.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <AppConfig.h>
+#include <AppTask.h>
+#include <AppTimer.h>
+
+AppTimer::AppTimer() : mTag("*"), mIsActive(false), mContext(nullptr), mHandler(nullptr), mCallback(nullptr) {}
```
Comment (mspang) [inline examples/window-app/efr32/src/AppTimer.cpp:24]: nit: move initializers to class definition and default this ctor

File: examples/window-app/efr32/src/WindowCover.cpp [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+#include <AppTask.h>
+#include <WindowCover.h>
+
+#define LIFT_DELTA (LCD_COVER_SIZE / 10)
+#define TILT_DELTA 1
+#define TIMER_DELAY_MS 500
+#define STATUS_DEFAULT 0x03 // bit0: Operational, bit1: Online
+
+WindowCover::WindowCover() :
+    mStatus(STATUS_DEFAULT), mType(kType_Tilt_Lift_blind), mLiftOpenLimit(LIFT_OPEN_LIMIT), mLiftClosedLimit(LIFT_CLOSED_LIMIT),
```
Comment (mspang) [inline examples/window-app/efr32/src/WindowCover.cpp:30]: nit: move default ctor initializers to class definition

File: examples/window-app/efr32/src/ZclCallbacks.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+/** @brief Window Covering Cluster Init
+ *
+ * Cluster Init
+ *
+ * @param endpoint    Endpoint that is being initialized
+ */
+void emberAfWindowCoveringClusterInitCallback(chip::EndpointId endpoint)
+{
+    EFR32_LOG("Window Covering Cluster init");
+    return;
```
Comment (mspang) [inline examples/window-app/efr32/src/ZclCallbacks.cpp:51]: nit: remove this return

File: examples/window-app/efr32/src/lcd.c [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+#define COVERING_TYPE_PROJECTOR_SCREEN 9             // Tilt
+
+#define BAND_COUNT 12
+#define BAND_SIZE (LCD_COVER_SIZE / BAND_COUNT)
+
+extern GLIB_Context_t glibContext;
+
+typedef int8_t (*LCDCoverPainter)(uint32_t x, uint32_t y, uint8_t lift, uint8_t tilt);
+
+uint32_t sTiltIcon[] = {
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:43]: nit: Declare this constexpr (or at least const)

File: examples/window-app/efr32/src/lcd.c [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    0xffffffff, 0xffffffff, 0xc0000003, 0xc0000003, 0xc0000003, 0xc0000003, 0xc1ffffc3, 0xc1ffffc3,
+    0xc1ffffc3, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003,
+    0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003, 0xc003c003,
+    0xc003c003, 0xc003c003, 0xc0000003, 0xc0000003, 0xc0000003, 0xc0000003, 0xffffffff, 0xffffffff,
+};
+
+uint32_t sLiftIcon[] = {
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:50]: nit: also const

File: examples/window-app/efr32/src/lcd.c [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            return (x < (LCD_MARGIN_SIZE + LCD_BORDER_SIZE)) || (x >= (LCD_SIZE - LCD_BORDER_SIZE - LCD_MARGIN_SIZE)) ||
+                (y < (LCD_MARGIN_SIZE + LCD_BORDER_SIZE)) || (y >= (LCD_SIZE - LCD_BORDER_SIZE - LCD_MARGIN_SIZE));
+        }
+    }
+    return -1;
+}
+
+static int8_t LCDIconPainter(uint32_t x, uint32_t y, uint8_t icon_id)
+{
+    uint32_t * icon = 0;
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:79]: nit: 0 -> nullptr

and const

File: examples/window-app/efr32/src/lcd.c [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+ *    limitations under the License.
+ */
+
+#include <AppConfig.h>
+#include <display.h>
+#include <dmd.h>
+#include <glib.h>
+#include <lcd_cover.h>
+
+#define COVERING_TYPE_ROLLERSHADE 0                  // Lift
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:25]: Can we use an enum cass for COVERING_TYPE ?

```
enum class CoveringType {
  kRollershade = 0,
  kDrapery = 1,
  // ...
};
```

File: examples/window-app/efr32/src/lcd.c [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+#define COVERING_TYPE_ROLLERSHADE_EXTERIOR 2         // Lift
+#define COVERING_TYPE_ROLLERSHADE_EXTERIOR_2_MOTOR 3 // Lift
+#define COVERING_TYPE_DRAPERY 4                      // Lift
+#define COVERING_TYPE_AWNING 5                       // Lift
+#define COVERING_TYPE_SHUTTER 6                      // Tilt
+#define COVERING_TYPE_TILT_BLIND 7                   // Tilt
+#define COVERING_TYPE_TILT_LIFT_BLIND 8              // Lift, Tilt
+#define COVERING_TYPE_PROJECTOR_SCREEN 9             // Tilt
+
+#define BAND_COUNT 12
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:36]: nit: #define -> constexpr uint32_t

(and BAND_SIZE as well)

File: examples/window-app/efr32/src/lcd.c [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+#define BAND_SIZE (LCD_COVER_SIZE / BAND_COUNT)
+
+extern GLIB_Context_t glibContext;
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:39]: nit: Please move declaration to a header file that is included by the file that defines this variable.

File: examples/window-app/efr32/src/lcd.c [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+typedef int8_t (*LCDCoverPainter)(uint32_t x, uint32_t y, uint8_t lift, uint8_t tilt);
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:41]: nit: prefer using to typedef

`using LCDCoverPainter = int8_t (*)(uint32_t x, uint32_t y, uint8_t lift, uint8_t tilt);`

File: examples/window-app/efr32/src/lcd.c [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+};
+
+uint32_t sLiftIcon[] = {
+    0xffffffff, 0xffffffff, 0xc0000003, 0xc0000003, 0xc0000003, 0xc0000003, 0xc0f00003, 0xc0f00003,
+    0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003,
+    0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0f00003, 0xc0ffffc3,
+    0xc0ffffc3, 0xc0ffffc3, 0xc0000003, 0xc0000003, 0xc0000003, 0xc0000003, 0xffffffff, 0xffffffff,
+};
+
+#define ICON_OFFSET 48
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:57]: nit: Prefer contexpr variables in an unnamed namespace to macros

File: examples/window-app/efr32/src/lcd.c [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+#define ICON_SIZE 32
+
+static int8_t LCDFramePainter(uint32_t x, uint32_t y)
```
Comment (mspang) [inline examples/window-app/efr32/src/lcd.c:60]: nit: Prefer unnamed namespace to `static`

(and many more below)

**Comment by jmartinez-silabs** [comment]
PTAL @andreilitvin @mspang @woody-apple. 
@ricardo-casallas made the requested changes. Could we get this merged ?

**Review by mspang — APPROVED** [review]

**Pull request merged**